### ISSUE TITLE
Try integrating into Create React App

### ISSUE THREAD

**Issue by gaearon** [pmmmwh/react-refresh-webpack-plugin#7]
I've tried integrating this into an ejected CRA project but got stuck (https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/6).

This gets me thinking: perhaps, the next milestone is to get it to a point where we can create a pull request for CRA? What's missing for that to work? cc @iansu

**Comment by iansu** [comment]
I would love to get this into Create React App. If there's anything needed on our end (other than installing and using the plugin) please let me know.

**Comment by Ericnr** [comment]
Would be great to see this being implemented without ejecting from CRA too, using Craco or React-App-Rewired.

**Comment by gaearon** [comment]
Yes, that’s what I’m referring to by “creating a pull request for CRA” 🙂 that was the goal all along.

**Comment by pmmmwh** [comment]
> I've tried integrating this into an ejected CRA project but got stuck (#6).
> 
> This gets me thinking: perhaps, the next milestone is to get it to a point where we can create a pull request for CRA? What's missing for that to work? cc @iansu

To my knowledge, most of the work have to be done here:
1. Allow custom error overlay and alter react error overlay/the API needed here
2. Fix assumptions when sniffing Babel/Webpack, since CRA does not use webpack's mode argument
3. Need to try come up with an integration with the Webpack socket situation (related #9 ), and also make this work with servers under proxy/custom paths)

**Comment by gaearon** [comment]
Could Babel detection be failing because the injected module doesn’t have a path that matches files going though Babel? Like usually people limit it to their source directory.

**Comment by pmmmwh** [comment]
> Could Babel detection be failing because the injected module doesn’t have a path that matches files going though Babel? Like usually people limit it to their source directory.

Yes exactly. People often ignore `node_modules` so the injected module does not get processed, thus yielding high percentage of false positives.

**Comment by esetnik** [comment]
Based on [the suggestion by @drather19](https://github.com/facebook/react/issues/16604#issuecomment-561961608) I have published a customize-cra plugin to make it easier. See [esetnik/customize-cra-react-refresh](https://github.com/esetnik/customize-cra-react-refresh).

**Comment by bugzpodder** [comment]
I tested this with mostly a CRA 3.3.0 build and got a few issues:
1. Got the same error in #29 and that fix worked for me.
2. The first change always refreshes the page for some reason.  subsequent changes doesn't.
3. After a bit I get "The development server has disconnected.
Refresh the page if necessary. "  then if I make a change nothing happens (page doesn't refresh/update).

**Comment by jimmyn** [comment]
I'm using [craco](https://github.com/gsoft-inc/craco) with configurations from this [codesandbox example](https://codesandbox.io/s/craco-fast-refresh-fzuj3?from-embed) and it seems to be working fine, but I'm also fighting with duplicated error overlays - https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/28

**Comment by charrondev** [comment]
I have it working perfectly in CRA as far as I know. I did the following:

- Start a CRA.
- Eject.
- Add the babel plugin.
- Add the webpack plugin w/ `disableRefreshCheck: true`
  ```js
      isEnvDevelopment && new ReactRefreshWebpackPlugin({ disableRefreshCheck: true }),
  ```
- Change the entry to use the default dev server instead of the CRA one.
  ```js
      // Note: instead of the default WebpackDevServer client, we use a custom one
      // to bring better experience for Create React App users. You can replace
      // the line below with these two lines if you prefer the stock client:
      // require.resolve('webpack-dev-server/client') + '?/',
      // require.resolve('webpack/hot/dev-server'),
      isEnvDevelopment &&
        // require.resolve('react-dev-utils/webpackHotDevClient'),
        require.resolve('webpack/hot/dev-server'),
  ```

I have no overlay issues or warnings. I haven't experienced the other issues described by @bugzpodder.

**Comment by charrondev** [comment]
After playing with it a little bit, I did a quick fork of the `react-dev-utils/webpackHotDevClient`. This removes all the error & warning handling.

https://gist.github.com/charrondev/c06cd5a9616a1d1a0db45de6f10b592b#file-webpackhotdevclient-js

If this is used instead of the `react-dev-utils` one I find

- There are no duplicate errors.
- Errors properly dismiss after being fixed.

**Comment by pmmmwh** [comment]
facebook/create-react-app#8582 has been merged, but some API naming work still has to be done so I'll leave this open.

**Comment by thoamsy** [comment]
> I'm using [craco](https://github.com/gsoft-inc/craco) with configurations from this [codesandbox example](https://codesandbox.io/s/craco-fast-refresh-fzuj3?from-embed) and it seems to be working fine, but I'm also fighting with duplicated error overlays - #28

I also use craco, but I am get the another error。The console says
> Uncaught ReferenceError: $RefreshReg$ is not defined
    at Object../node_modules/events/events.js (events.js:447)
    at __webpack_require__ (bootstrap:806)
    at fn (bootstrap:167)
    at Object../node_modules/webpack/hot/emitter.js (emitter.js:1)
    at __webpack_require__ (bootstrap:806)

**Comment by pmmmwh** [comment]
> I also use craco, but I am get the another error

Did you include the Babel plugin in your config? I won't be able to help unless there is a reproducible example.

**Comment by thoamsy** [comment]
> Did you include the Babel plugin in your config? I won't be able to help unless there is a reproducible example.

Thank you, I will try to provide a example for you.

**Comment by thoamsy** [comment]
It ’s weird, my project will report an error, but when I wrote a minimal demo using craco, it worked very well.

**Comment by bugzpodder** [comment]
> It ’s weird, my project will report an error, but when I wrote a minimal demo using craco, it worked very well.

do you have an example where i can test as well?  i recently had tried to apply to a project but couldn't get it to work.

**Comment by thoamsy** [comment]
> do you have an example where i can test as well? i recently had tried to apply to a project but couldn't get it to work.
This is [my demo](https://codesandbox.io/s/twilight-sound-ipp98) for test the hot reload, it works very well. But I can't bring it to my project.

**Comment by bugzpodder** [comment]
@thoamsy craco didn't work for me (got $RefreshReg$ error), however I tried https://dutzi.party/react-fast-refresh/ and it worked better

**Comment by pmmmwh** [comment]
@thoamsy @bugzpodder 
I'm not sure what broke for you guys, but I am able to get the plugin working with `craco`. [Here](https://codesandbox.io/s/using-craco-ei2ek) is a CodeSandbox demonstrating this.

I won't be able to help without any more insights into the specific setups for your apps, or with more debugging information.

**Comment by thoamsy** [comment]
@bugzpodder Do you mean that use `react-app-rewired` instead of `craco` can works well in your project?

**Comment by thoamsy** [comment]
> @thoamsy @bugzpodder
> I'm not sure what broke for you guys, but I am able to get the plugin working with `craco`. [Here](https://codesandbox.io/s/using-craco-ei2ek) is a CodeSandbox demonstrating this (Notice the use of `whenDev`, that is important because the plugin do not inject required runtime code in non-dev modes, and will cause the Babel plugin to break builds because of missing globals).

Thank for your advice, but I think our problem is that the dev mode can't run 😂。Maybe I should try to use another tool instead of `craco` to let my project works 🤔

**Comment by bugzpodder** [comment]
@thoamsy yes, but react-app-rewired just modifies your webpack config like craco so you can get it to work with some modification I believe.

@pmmmwh that's because you are relying on codesandbox.  I am doing it locally on a mac.  Just goto your codesandbox, download it as ZIP, install it locally, add your plugin (0.3.0-beta.5) and yarn start it.  I always only used `yarn start` so whenDev didn't really matter.

<img width="515" alt="Screen Shot 2020-04-26 at 9 01 30 AM" src="https://user-images.githubusercontent.com/14841421/80312959-00210280-879d-11ea-8955-fc9ecf99af1b.png">

**Comment by pmmmwh** [comment]
@thoamsy @bugzpodder
So I did some debugging, it seems that the issue here is related to gsoft-inc/craco#36 and gsoft-inc/craco#49 .

The problem is that this plugin by default don't process files within `node_modules`, but `craco` will apply `react-refresh/babel` unconditionally to both instances of `babel-loader`, thus breaking all JS files within `node_modules`.

To solve this, other than requesting a proper fix on `craco`'s side, you can manually inject the Babel plugin as follows:

```js
module.exports = {
  webpack: {
    configure: (webpackConfig) => {
      const babelLoaderRule = webpackConfig.module.rules
        .find(x => Array.isArray(x.oneOf))
        .oneOf.find(
          x =>
            x.loader &&
            x.loader.includes('babel-loader') &&
            x.options &&
            x.options.presets &&
            x.options.presets.some(p => p.includes('babel-preset-react-app'))
        );

      babelLoaderRule.options.plugins.push(require.resolve('react-refresh/babel'));

      return webpackConfig;
    },
    plugins: [
      ...whenDev(
        () => [
          new ReactRefreshWebpackPlugin({
            // Proper setup for this would require the new unpublished version of CRA.
            // We'll disable the integration for now.
            overlay: false,
          }),
        ],
        []
      ),
    ],
  }
}
```

**Comment by thoamsy** [comment]
@pmmmwh It works! Thank you for your effort to solve this problem, it's amazing!

**Comment by kelly-tock** [comment]
Is this released already with CRA? I don’t see the env variable mentioned in the PR in the documentation

**Comment by deftomat** [comment]
@kelly-tock Yes, it is. Use `FAST_REFRESH` env variable.

**Comment by gaearon** [comment]
Note there are some pending fixes that haven't made into CRA version yet.

**Comment by pmmmwh** [comment]
> Is this released already with CRA?

CRA haven't had a release after the PR was merged, so it's not out yet.

**Comment by kelly-tock** [comment]
@pmmmwh that's exactly what I was asking. thanks!

**Comment by xyy94813** [comment]
I am using `react-hot-loader` with CRA.
But i got some problems when i use code splitting.
So, i am looking for a new solution.

If i want to use it with CRA now.
Is it the best option to use `esetnik/customize-cra-react-refresh` and then remove it after the CRA published?

**Comment by pmmmwh** [comment]
> Is it the best option to use `esetnik/customize-cra-react-refresh` and then remove it after the CRA published?

It depends on your setup, e.g. if you're using any rewire tools. I would suggest you take a look at the `examples/create-react-app` folder, then if you really don't want to bother you can move on with the rewire plugin.

**Comment by Saibamen** [comment]
Fixed - released in CRA 4.0.0
